Enable running projects without MLproject or conda.yaml files

smurching opened the pull request:
Enables running projects (git repositories or local directories) that don't contain MLproject or conda.yaml files. This is useful e.g. for running code / scripts that are unaware of MLflow via `mlflow run` (note: ideally we'd also detect / autoinstall requirements.txt, but this PR doesn't add that functionality).

smurching committed:
Check that conda_env is specified in Project YAML before attempting to use it

smurching committed:
Revert "Check that conda_env is specified in Project YAML before attempting to use it"

This reverts commit 7a232dce6952c07c02050c906a86e349d5f80a25.

smurching committed:
update

mateiz commented:
If you create an empty conda environment, don't make the name based on the project; that would mean recreating it every time they run. Just have one empty one based on the hash of the empty string or something like that.

smurching committed:
WIP

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -57,6 +57,10 @@ def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=N
+            if not project.conda_env:
+                raise ExecutionException(
+                    "MLproject file must have a 'conda_env' field containing the path to a "
+                    "YAML conda environment file.")
```
"or `--no-conda` should be provided"

smurching committed:
Merge branch 'master' into fix-require-conda-env

smurching committed:
Fix typo / test (os.abspath -> os.path.abspath)

mateiz commented on mlflow/projects/__init__.py:
```diff
@@ -258,10 +257,20 @@ def _fetch_git_repo(uri, version, dst_dir, git_username, git_password):
+    conda_env_contents = ""
+    if project.conda_env:
+        with open(os.path.join(project_dir, project.conda_env)) as conda_env_file:
+            conda_env_contents = conda_env_file.read()
```
Does this work if the project does not declare a `conda_env` but does contain a `conda.yaml`? We want that to work too. We don't want to force users to create an MLproject file.

smurching replied:
Ah gotcha, it doesn't work currently in the case you described - will fix it

smurching replied:
To clarify - we do need an MLproject file because it contains the project definition (entry points etc) right? But we'll default to looking for conda.yaml in the same directory as the MLproject file if no `conda_env` is specified

smurching replied:
Actually - since we allow running entry points not specified in MLproject files I guess we don't need to require the MLproject either (we currently do)

smurching committed:
WIP enabling running projects that don't contain an MLproject file

smurching committed:
WIP, added new test for load_project

smurching committed:
Update load_projects test

smurching committed:
Fix linter error (repeated import)

smurching committed:
Merge master

smurching committed:
Fix test failure (passing string instead of Project to _get_entry_point)

aarondav commented:
@smurching Please rebase - let's get this guy in!

smurching committed:
Merge master

smurching committed:
Merge branch 'master' into fix-require-conda-env

smurching committed:
Fix tests after merging master; needed to use os.path.abspath in _fetch_project

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -283,11 +267,15 @@ def _maybe_create_conda_env(conda_env_path):
-        eprint('=== Creating conda environment %s ===' % conda_env)
-        process.exec_cmd([conda_path, "env", conda_action, "-n", conda_env, "--file",
-                          conda_env_path], stream_output=True)
+    project = _project_spec.load_project(project_dir)
```
It seems like this method should take a project, instead of project_dir, since its caller has one.

smurching replied:
Oops my mistake :P

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -268,8 +253,7 @@ def _get_conda_bin_executable(executable_name):
 
-def _maybe_create_conda_env(conda_env_path):
-    conda_env = _get_conda_env_name(conda_env_path)
+def _maybe_create_conda_env(project_dir):
```
Can we add documentation to this method as to what the `maybe` means (i.e., if not created before). Perhaps if this were `_get_or_create_conda_env`, and it returned the conda environment name, which would be passed into `_get_entry_point_command` instead of `use_conda`?

aarondav commented on mlflow/projects/_project_spec.py:
```diff
@@ -1,25 +1,44 @@
+
+
+def load_project(directory):
+    mlproject_path = os.path.join(directory, "MLproject")
```
Let's make environment variables with MLproject and conda.yaml, as these are important default constants.

aarondav commented on mlflow/projects/_project_spec.py:
```diff
@@ -1,25 +1,44 @@
+        command = entry_point_yaml.get("command")
+        entry_points[name] = EntryPoint(name, parameters, command)
+    conda_path = yaml_obj.get("conda_env")
+    if conda_path:
```
What happens if conda_path is specified, but the file does not exist? What kind of error do we get?

smurching replied:
Good question, we ultimately get [an exception](https://github.com/mlflow/mlflow/pull/218/files#diff-16104674981d54b31e320402a143eb19R201) when `load_conda_env` is called. We could validate here, although we might still want to validate when loading the environment e.g. in case conda.yaml is deleted (or we could allow for an uglier error message in that case, since it seems like an edge case). `conda_path` itself is also used directly when running `conda env create --file <conda_path>` to create the environment.

smurching replied:
I'll add a check verifying that `conda_path` points to an existing file here (during project-loading) for now.

aarondav commented on mlflow/projects/_project_spec.py:
```diff
@@ -1,25 +1,44 @@
-        # TODO: validate the spec, e.g. make sure paths in it are fine
+    """A project specification loaded from an MLproject file in the passed-in directory."""
+    def __init__(self, conda_env_path, entry_points):
+        self.conda_env_path = conda_env_path
```
conda_env_path should be able to be private since we provide `load_conda_env` -- if true, we should name this variable `self._conda_env_path` to help indicate that it should not be accessed outside of the class.

smurching replied:
Good question, we actually do use `conda_env_path` when running `conda env create --file <conda_path>` to create the conda environment, so I've kept it public for now

aarondav commented on mlflow/projects/_project_spec.py:
```diff
@@ -1,25 +1,44 @@
+    """A project specification loaded from an MLproject file in the passed-in directory."""
+    def __init__(self, conda_env_path, entry_points):
+        self.conda_env_path = conda_env_path
+        self.entry_points = entry_points
```
Potentially same w/entry_points since we have `get_entry_point`

aarondav commented on mlflow/projects/__init__.py:
```diff
@@ -299,16 +287,14 @@ def _maybe_set_run_terminated(active_run, status):
-        conda_env_path = os.path.abspath(os.path.join(project_dir, project.conda_env))
         activate_path = _get_conda_bin_executable("activate")
-        commands.append("source %s %s" % (activate_path, _get_conda_env_name(conda_env_path)))
+        commands.append("source %s %s" % (activate_path, _get_conda_env_name(project)))
```
Is there any unit test which confirms that we can successfully run through this code with an absent conda env, but still create and activate the empty conda environment? (That is the behavior we expect, right?)

smurching committed:
Respond to some review comments; refactor constants & conda env creation logic

smurching committed:
Add integration test verifying that new conda environment works properly

smurching committed:
Fix test (forgot to refactor after making entry_points private in Projects). Also add logic for
handling unicode output from shlex_quote

aarondav commented:
LGTM!

smurching committed:
Deubgging experiment: run tests in conda environment instead of in virtualenv with tox

smurching committed:
Use quiet flag for pip installation

smurching committed:
Debug: only run projects tests

smurching committed:
Debug: move quiet option to end of pip install command

smurching committed:
Debug: run all tests

smurching committed:
Remove CLI integration test in favor of a large unit test that
verifies that our conda-env-creation method behaves as expected given a project without a conda env

smurching committed:
Undo changes to travis

smurching committed:
Revert "Remove CLI integration test in favor of a large unit test that"

This reverts commit e88385ff29d808576cfa0d580cd63cf21d963825.

smurching committed:
Debug: Pass expected conda env name to script in integration test

smurching committed:
Merge branch 'master' into fix-require-conda-env

smurching committed:
Update assertion error message in test for clarity
